// @feathersjs/authentication-oauth/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = __importDefault(require("debug"));
const lodash_1 = require("lodash");
const strategy_1 = require("./strategy");
exports.OAuthStrategy = strategy_1.OAuthStrategy;
const express_1 = __importDefault(require("./express"));
const utils_1 = require("./utils");
const debug = debug_1.default('@feathersjs/authentication-oauth');
exports.setup = (options) => (app) => {
    const service = app.defaultAuthentication ? app.defaultAuthentication(options.authService) : null;
    if (!service) {
        throw new Error('An authentication service must exist before registering @feathersjs/authentication-oauth');
    }
    const { oauth } = service.configuration;
    if (!oauth) {
        debug(`No oauth configuration found in authentication configuration. Skipping oAuth setup.`);
        return;
    }
    const { strategyNames } = service;
    // Set up all the defaults
    const { path = '/oauth' } = oauth.defaults || {};
    const port = app.get('port');
    let host = app.get('host');
    let protocol = 'https';
    // Development environments commonly run on HTTP with an extended port
    if (app.get('env') === 'development') {
        protocol = 'http';
        if (String(port) !== '80') {
            host += ':' + port;
        }
    }
    const grant = lodash_1.merge({
        defaults: {
            path,
            host,
            protocol,
            transport: 'session'
        }
    }, lodash_1.omit(oauth, 'redirect'));
    const getUrl = (url) => {
        const { defaults } = grant;
        return `${defaults.protocol}://${defaults.host}${path}/${url}`;
    };
    lodash_1.each(grant, (value, name) => {
        if (name !== 'defaults') {
            value.callback = value.callback || getUrl(`${name}/authenticate`);
            value.redirect_uri = value.redirect_uri || getUrl(`${name}/callback`);
            if (!strategyNames.includes(name)) {
                debug(`Registering oAuth default strategy for '${name}'`);
                service.register(name, new strategy_1.OAuthStrategy());
            }
        }
    });
    app.set('grant', grant);
};
exports.express = (settings = {}) => (app) => {
    const options = utils_1.getDefaultSettings(app, settings);
    app.configure(exports.setup(options));
    app.configure(express_1.default(options));
};
exports.expressOauth = exports.express;
